examples/stateful/stateful_handler.py

```diff
@@ -0,0 +1,79 @@
+import logging
+from abc import ABC
+from typing import Dict
+
+from lru import LRU
+
+from ts.context import Context
+from ts.torch_handler.base_handler import BaseHandler
+
+logger = logging.getLogger(__name__)
+
+
+class StatefulHandler(BaseHandler, ABC):
+    def __init__(self):
+        super().__init__()
+        self.cache: LRU = None
+        self.sequence_ids: Dict = None
+        self.context = None
+
+    def initialize(self, ctx: Context):
+        """
+        Loads the model and Initializes the necessary artifacts
+        """
+
+        super().initialize(ctx)
+        self.context = ctx
+        if self.context.model_yaml_config["handler"] is not None:
+            self.cache = LRU(
+                int(self.context.model_yaml_config["handler"]["cache"]["capacity"])
+            )
+
+        self.initialized = True
+
+    def preprocess(self, data):
+        """
+        Preprocess function to convert the request input to a tensor(Torchserve supported format).
+        The user needs to override to customize the pre-processing
+
+        Args :
+            data (list): List of the data from the request input.
+
+        Returns:
+            tensor: Returns the tensor data of the input
+        """
+
+        self.sequence_ids = {}
+        results = []
+        for idx, row in enumerate(data):
```

Comment:
To confirm, is it the case that `batchSize` is the least upper bound of `len(data)`, i.e.`len(data) <= batchSize` and for all `l` such that `len(data) <= l`, `batchSize <= l`?

Is it possible for two separate requests to get batched to this worker? If so, suppose there are two separate streaming requests that are batched to this worker. What happens if one client is much much faster than the other? Do we throttle the faster client to match the speed of the slower one by buffering the faster client's messages?

Reply:
- Q1: yes, len(data) <= batchSize. data is a batch of request received at realtime.

- Q2: Yes, a batch of requests comes from different sequences. eg. len(data) = 4, it means there are 4 sequences. Each sequence has its own dedicated jobQ. Only the parameter "maxBatchDelay" decides the msec of batching a group of requests from different sequences. In other words, the different traffic volume of different sequences has no impact on batching latency.

Reply:
Ok but if two streams produce data at drastically different rates, how do you keep the batch index coherent? For instance, fix a stateful worker.  At time `t_0`, the worker receives data `d_0_0` and `d_1_0` from two streams. So then `len(data) == 2` and `data[0]` is the payload for stream 0 and `data[1]` is the payload for stream 1. 

At `t_1`, stream 0 does not produce any data because it took longer than `maxBatchDelay`, but stream 1 produces data `d_1_1`. So then `len(data) == 1` and `data[0]` is the payload for stream 1. In the line below, `idx == 0`, so then you fetch the sequence ID for index 0. It seems like this would fetch the sequence ID for stream 0, 
```python
sequence_id = self.context.get_sequence_id(idx)
```
but you actually want the sequence ID for stream 1. Am I understanding the API semantics correctly?

Reply:
each request's sequence id is added into its header with key = "ts_request_sequence_id". Backend can get a request's sequence id via its header. This can guarantee we can always get the sequence id regardless the real batch size is changed or the request of a sequence enters into a different batch slot.